UI/CombatReport/CombatLogWnd.cpp

```diff
@@ -66,52 +66,145 @@ class CombatLogWnd::Impl {
-        return objects_per_owner;
+
+        return forces_by_owner;
     }
 
-    // converts input empire ids and numbers of objects to text with names and
-    // links to empires and numbers-as-text of how many objects they own
-    std::string CountsToText(const std::map<int, int>& count_per_empire,
-                             const std::string& delimiter = ", ")
-    {
+    std::string EmpireIdToText(int empire_id) {
+        if (const Empire* empire = GetEmpire(empire_id))
+            return GG::RgbaTag(empire->Color()) + "<" + VarText::EMPIRE_ID_TAG + " " +
+                   std::to_string(empire->EmpireID()) + ">" + empire->Name() + "</" +
+                   VarText::EMPIRE_ID_TAG + ">" + "</rgba>";
+        else
+            return GG::RgbaTag(ClientUI::DefaultLinkColor()) + UserString("NEUTRAL") + "</rgba>";
+    }
+
+    /// converts to "Empire_name: n" text
+    std::string CountToText(int empire_id, int forces_count) {
         std::stringstream ss;
-        for (auto it = count_per_empire.begin(); it != count_per_empire.end(); ) {
-            std::string owner_string;
-            if (const Empire* owner = GetEmpire(it->first))
-                owner_string = GG::RgbaTag(owner->Color()) + "<" + VarText::EMPIRE_ID_TAG + " "
-                    + std::to_string(owner->EmpireID()) + ">" + owner->Name()
-                    + "</" + VarText::EMPIRE_ID_TAG + ">" + "</rgba>";
-            else
-                owner_string = GG::RgbaTag(ClientUI::DefaultLinkColor()) + UserString("NEUTRAL")  + "</rgba>";
-            ss << owner_string << ": " << it->second;
-            ++it;
-            if (it != count_per_empire.end())
-                ss << delimiter;
+        ss << EmpireIdToText(empire_id) << ": " << forces_count;
+        return ss.str();
+    }
+
+    /// Used to sort empire foces for display
+    class ForcesSorter {
+    public:
+        ForcesSorter(int viewing_empire_id_)
+            : viewing_empire_id(viewing_empire_id_) {
+            }
+                    
+        bool operator()(const std::shared_ptr<UniverseObject>& lhs,
+                        const std::shared_ptr<UniverseObject>& rhs) {
+            if (TypeComparison(lhs->ObjectType(), rhs))
+                return true;
+
+            if (TypeComparison(rhs->ObjectType(), lhs))
+                return false;
+
+            const std::string lhs_public_name = lhs->PublicName(viewing_empire_id);
+            const std::string rhs_public_name = rhs->PublicName(viewing_empire_id);
+            if (lhs_public_name < rhs_public_name)
+                return true;
+
+            if (lhs_public_name > rhs_public_name)
+                return false;
+
+            return lhs->ID() < rhs->ID();
+        }
+
+        static bool TypeComparison(UniverseObjectType type,
+                                   const std::shared_ptr<UniverseObject>& object) {
+            return type < object->ObjectType();
+        }
+
+    private:
+        int viewing_empire_id;
```

Comment:
should default initilalize with `ALL_EMPIRES`

Reply:
You mean the constructor parameter? There's no point in defaulting the field if there's no default constructor.

Reply:
I should make the constructor explicit, while I'm at it.